src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PreReleaseUserService.cs

```diff
@@ -153,6 +158,93 @@ public class PreReleaseUserService : IPreReleaseUserService
                 );
         }
 
+        public async Task<Either<ActionResult, Unit>> RemovePreReleaseUser(Guid releaseId, string email)
+        {
+            if (!new EmailAddressAttribute().IsValid(email))
+            {
+                return ValidationActionResult(InvalidEmailAddress);
+            }
+
+            return await _persistenceHelper
+                .CheckEntityExists<Release>(releaseId)
+                .OnSuccess(_userService.CheckCanAssignPrereleaseContactsToRelease)
+                .OnSuccessVoid(
+                    async () =>
+                    {
+                        _context.RemoveRange(
+                            _context
+                                .UserReleaseRoles
+                                .Include(r => r.User)
+                                .Where(
+                                    r =>
+                                        r.ReleaseId == releaseId
+                                        && r.User.Email.ToLower() == email.ToLower()
+                                        && r.Role == ReleaseRole.PrereleaseViewer
+                                )
+                        );
+                        _context.RemoveRange(
+                            _context
+                                .UserReleaseInvites
+                                .Where(
+                                    i =>
+                                        i.ReleaseId == releaseId
+                                        && i.Email.ToLower() == email.ToLower()
+                                        && i.Role == ReleaseRole.PrereleaseViewer
+                                )
+                        );
+                        await _context.SaveChangesAsync();
+
+                        var remainingReleaseInvites = await _context
```

Comment:
A little code comment explanation here might be helpful.  So we're saying that if this user has no more UserReleaseInvites invites for the "PreReleaseViewer" role against any Release, then remove any "UserInvites" record for them as well from the UsersAndRolesDbContext which would allow them to be assigned the global PreRelease user role.

Reply:
This isn't actually my code - it's been highlighted in git because I moved the private methods to the bottom.

I would expect for this code to remove the UserInvites row for that specific release in the same way it removes it in the UserReleaseInvites table. It seems that currently, if a user has other prerelease invites in the UserReleaseInvites table, unrelated to the release in question, then the associated UserInvites row will *not* be removed. Not sure what this means if that "removed" user then tries to log in to the service.

Since this is unrelated to my changes I might make a separate ticket to investigate this. You happy with that?

Reply:
Yup!  I suspect it's doing the right thing, preventing the user from being given an AspNetUserRoles entry if they have no more PreRelease invites for any Releases.

Reply:
Okay, I understand it now. UserInvites only record whether they have a specific role - it doesn't record what release that role is against. I'll add a comment.